### ISSUE TITLE
get ssl error

### ISSUE THREAD

**Issue by leetomlee123** [elazarl/goproxy#600]
2024/12/25 15:11:20 [007] INFO: on 0th handler: &{2 <nil> 0x1220e00} 47.243.34.160:80
2024/12/25 15:11:20 [007] INFO: Assuming CONNECT is TLS, mitm proxying it
2024/12/25 15:11:20 [007] INFO: signing for 47.243.34.160

2024/12/25 15:11:20 [007] WARN: Cannot handshake client 47.243.34.160:80 tls: first record does not look like a TLS handshake

data stream=>  android  vpnservice->tun2proxy->goproxy 

error forward http request to https request 

how to fix it

**Comment by ErikPelli** [comment]
There are so many middle services so it's hard to follow.
By the way, I assume you are currently indirectly using the `MitmConnect` action, through goproxy.AlwaysMitm.
You have to handle to add a different handler for the port 80 (HTTP) to use the `HTTPMitmConnect`, for normal HTTP data.

In goproxy, http mitm and https are separated.

**Comment by leetomlee123** [comment]
`

	customCaMitm := &goproxy.ConnectAction{Action: goproxy.ConnectMitm, TLSConfig: goproxy.TLSConfigFromCA(cert)}
	customCaMitmHttp := &goproxy.ConnectAction{Action: goproxy.ConnectHTTPMitm, TLSConfig: goproxy.TLSConfigFromCA(cert)}
    var customAlwaysMitm goproxy.FuncHttpsHandler = func(host string, ctx *goproxy.ProxyCtx) (*goproxy.ConnectAction, string) {
        if strings.HasSuffix(host, ":80") || strings.Contains(host, ":80") {
            log.Printf("HTTP request intercepted: %s", host)
            return customCaMitmHttp, host
        } else if strings.HasSuffix(host, ":443") || strings.Contains(host, ":443") {
            log.Printf("HTTPS request intercepted: %s", host)
            return customCaMitm, host
        } else {
            log.Printf("Unknown request intercepted: %s", host)
            return customCaMitm, host
        }
    }
	proxy := goproxy.NewProxyHttpServer()
	proxy.Verbose = true
	proxy.CertStore = NewCertStorage()
	proxy.OnRequest().HandleConnect(customAlwaysMitm)

`
this is my code ,its can not get https ,code right ? help

**Comment by leetomlee123** [comment]
![4c3d595728368b23ae7c57552d9c140](https://github.com/user-attachments/assets/1a842a49-e065-4c56-b788-8013e137f2f2)

I can reproduce this problem by using v2rayng to set the http proxy to goproxy. There is no problem when I use wifi http proxy to connect to goproxy.

**Comment by ErikPelli** [comment]
If this problem happens only when using this project, I would analyze what it does and log the information about every single request.
I have no clue if they rewrite the port or what, never used it.
Also, your check doesn't seem to have any meaning, id you're inside handleConnect, the method is always CONNECT.
I think that your problem is that you're using a CONNECT proxy for a normal HTTP request, you need to investigate about this.

**Comment by leetomlee123** [comment]
customCaMitm := &goproxy.ConnectAction{Action: goproxy.ConnectMitm, TLSConfig: goproxy.TLSConfigFromCA(cert)}
	// customCaMitmHttp := &goproxy.ConnectAction{Action: goproxy.ConnectHTTPMitm, TLSConfig: goproxy.TLSConfigFromCA(cert)}
	
	var customAlwaysMitm goproxy.FuncHttpsHandler = func(host string, ctx *goproxy.ProxyCtx) (*goproxy.ConnectAction, string) {
			return customCaMitm, host // 使用 MITM
	}
	
	proxy := goproxy.NewProxyHttpServer()
	proxy.Verbose = true
	proxy.CertStore = NewCertStorage()
	proxy.OnRequest().HandleConnect(customAlwaysMitm)

thats my new code ,https req is ok,but http req failed ,how to  resolve it 

![image](https://github.com/user-attachments/assets/c58d0129-cb33-4982-a912-3ddbca19ffc7)

**Comment by leetomlee123** [comment]
As mentioned above, I use the tun2proxy library to forward all traffic to goproxy, but goproxy doesn't seem to be able to distinguish between http/https requests and processes them all as https. This error occurs. Can you tell me how to configure it to handle it correctly?

**Comment by leetomlee123** [comment]
I may have found the problem. tun2proxy is a VPN that will establish a connect tunnel between proxy servers to forward all traffic. So goproxy receives only connect requests, which leads to misjudgment and treats http as https.

**Comment by ErikPelli** [comment]
> I may have found the problem. tun2proxy is a VPN that will establish a connect tunnel between proxy servers to forward all traffic. So goproxy receives only connect requests, which leads to misjudgment and treats http as https.

Yes, this is what I told you in the last message.
You have to distinguish `ConnectHTTPMitm` and TLS MITM by yourself, maybe using the port.
The first is for HTTP, the latter for HTTPS.

If you use TLS for HTTP, obviously it doesn't work...

**Comment by leetomlee123** [comment]
var customAlwaysMitm goproxy.FuncHttpsHandler = func(host string, ctx *goproxy.ProxyCtx) (*goproxy.ConnectAction, string) {

		if ctx.Req.URL.Port() == "443" {
			return goproxy.MitmConnect, host
		} else if ctx.Req.URL.Port() == "80" {
			return goproxy.HTTPMitmConnect, host
		} else {
			return goproxy.RejectConnect, host
		}

	}

	proxy := goproxy.NewProxyHttpServer()
	proxy.Verbose = true
	proxy.CertStore = NewCertStorage()
	proxy.OnRequest().HandleConnect(customAlwaysMitm)
	
	
	This is my latest code. It can solve the problem of **_tls: first record does not look like a TLS handshake_**. It can capture http packets but not https packets.

**Comment by ErikPelli** [comment]
No updates about this issue from the original author, according to the messages it was likely an issue with the integration made by the user.
Closing this issue, if it's still present please let me know.
